### PR TITLE
Созонов Илья. Вариант 26. Технология TBB. Линейная фильтрация изображений (блочное разбиение). Ядро Гаусса 3x3. 

### PR DESCRIPTION
**Pull request by sozozzya**
**Стандартная свертка Гаусса (3×3):**

Ядро Гаусса 3x3 (нормированное) для σ=1 имеет вид: 

image

Каждому пикселю изображения ставится в соответствие новая величина, полученная как сумма произведений значений пикселей из окрестности 3×3 на коэффициенты ядра Гаусса.

Например, если центральный пиксель изображения **𝐼(𝑥, 𝑦)**, то его новое значение **𝐼′(𝑥, 𝑦)** будет: 

image

**Параллельная реализация:**

Для ускорения фильтрации изображений используется параллельная обработка блоков с помощью TBB.

**1. Разбиение на блоки**

- Задаётся размер блока `block_size` (например, 64×64).
- Изображение разбивается на прямоугольные блоки по координатам `i` и `j` с шагом `block_size`.

**2. Параллельная обработка блоков**

- Используется `tbb::task_group`, где каждая задача (`tg.run`) обрабатывает отдельный блок изображения.
- Внутри блока выполняется стандартная свёртка с Гаусса (3×3).
- Обработка избегает граничных пикселей, чтобы не выходить за пределы массива.

**3. Запись результата**

- Каждый поток записывает результат фильтрации в общий выходной массив.
- Потоки работают с разными блоками, поэтому синхронизация не требуется.

### CODE REVIEW

File: tasks/tbb/sozonov_i_image_filtering_block_partitioning/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    img[((hght - 1) * wdth) + i] = 0;
+  }
+  for (int i = 1; i < hght - 1; ++i) {
+    img[i * wdth] = 0;
+    img[(i * wdth) + wdth - 1] = 0;
+  }
+
+  return img;
+}
+
```
Comment (Ivan-Lysov) [inline tasks/tbb/sozonov_i_image_filtering_block_partitioning/func_tests/main.cpp:25]: add a test for an empty image

**Review by Ivan-Lysov — COMMENTED** [review]
add a test for an empty image

File: tasks/tbb/sozonov_i_image_filtering_block_partitioning/src/ops_tbb.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    }
+  }
+
+  // Check size of image
+  return task_data->inputs_count[0] > 0 && task_data->inputs_count[0] == img_size &&
+         task_data->outputs_count[0] == img_size && task_data->inputs_count[1] >= 3 && task_data->inputs_count[2] >= 3;
+}
+
+bool sozonov_i_image_filtering_block_partitioning_tbb::TestTaskTBB::RunImpl() {
+  int block_size = 32;
```
Comment (ascannel) [inline tasks/tbb/sozonov_i_image_filtering_block_partitioning/src/ops_tbb.cpp:73]: I suggest you to increase block size, I think it will increase performance for your particular task

**Review by ascannel — CHANGES_REQUESTED** [review]

**Review by Ivan-Lysov — APPROVED** [review]

**Review by ascannel — APPROVED** [review]

**Review by allnes — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]
